crypto/rand/rand_vms.c

```diff
@@ -28,111 +36,415 @@
- * bit of entropy per 2 data bits.
+ * For the moment, we assume about 0.05 entropy bits per data bit, or 1
+ * bit of entropy per 20 data bits.
  */
-#define ENTROPY_FACTOR  2
+#define ENTROPY_FACTOR  20
 
 size_t rand_pool_acquire_entropy(RAND_POOL *pool)
 {
-    /* determine the number of items in the JPI array */
-    struct items_data_st item_entry;
-    size_t item_entry_count = OSSL_NELEM(items_data);
-    /* Create the 32-bit JPI itemlist array to hold item_data content */
-    struct {
-        uint16_t length, code;
-        uint32_t *buffer;
-        uint32_t *retlen;
-    } item[item_entry_count], *pitem;
-    struct items_data_st *pitems_data;
-    /* 8 bytes (two longs) per entry max */
-    uint32_t data_buffer[(item_entry_count * 2) + 4];
-    uint32_t iosb[2];
-    uint32_t sys_time[2];
-    uint32_t *ptr;
-    size_t i, j ;
-    size_t tmp_length   = 0;
+    ILE3 JPI_items_64bit[OSSL_NELEM(JPI_item_data) + 1];
+    ILE3 RMI_items_64bit[OSSL_NELEM(RMI_item_data) + 1];
+    uint32_t data_buffer_64bit[OSSL_NELEM(JPI_item_data_64bit) * 2
+                               + OSSL_NELEM(RMI_item_data_64bit) * 2];
+    size_t total_length_64bit = 0;
+    ILE3 DVI_items[OSSL_NELEM(DVI_item_data) + 1];
+    ILE3 JPI_items[OSSL_NELEM(JPI_item_data) + 1];
+    ILE3 RMI_items[OSSL_NELEM(RMI_item_data) + 1];
+    ILE3 SYI_items[OSSL_NELEM(SYI_item_data) + 1];
+    uint32_t data_buffer[OSSL_NELEM(DVI_item_data)
+                         + OSSL_NELEM(JPI_item_data)
+                         + OSSL_NELEM(RMI_item_data)
+                         + OSSL_NELEM(SYI_item_data)
+                         + 4 /* For JPI$_FINALEXC */ ];
     size_t total_length = 0;
     size_t bytes_needed = rand_pool_bytes_needed(pool, ENTROPY_FACTOR);
     size_t bytes_remaining = rand_pool_bytes_remaining(pool);
 
-    /* Setup itemlist for GETJPI */
-    pitems_data = items_data;
-    for (pitem = item; pitems_data->length != 0; pitem++) {
-        pitem->length = pitems_data->length;
-        pitem->code   = pitems_data->code;
-        pitem->buffer = &data_buffer[total_length];
-        pitem->retlen = 0;
-        /* total_length is in longwords */
-        total_length += pitems_data->length / 4;
-        pitems_data++;
-    }
-    pitem->length = pitem->code = 0;
+    /* Take all the 64-bit items first, to avoid possible alignment problems */
+    total_length_64bit +=
+        prepare_item_list(JPI_item_data_64bit, OSSL_NELEM(JPI_item_data_64bit),
+                          JPI_items_64bit,
+                          &data_buffer_64bit[total_length_64bit]);
+    total_length_64bit +=
+        prepare_item_list(RMI_item_data_64bit, OSSL_NELEM(RMI_item_data_64bit),
+                          RMI_items_64bit,
+                          &data_buffer_64bit[total_length_64bit]);
+    /* Now the 32-bit items */
+    total_length += prepare_item_list(DVI_item_data, OSSL_NELEM(DVI_item_data),
+                                      DVI_items, &data_buffer[total_length]);
```

Comment:
Wait a sec! total_length is in bytes, so where would &data_buffer[total_length] point? There ought to be your crash. I mean maybe it wasn't about alignment after all...

Reply:
> maybe it wasn't about alignment after all...

But if it is, then data_buffer_64bit still needs alignment *insurance*.

Reply:
It seems that insurance is still needed